File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_Android/JavaWrapperDemo/wrapper/src/main/java/com/hyphenate/wrapper/EMUserInfoManagerWrapper.java
package com.hyphenate.wrapper;

import com.hyphenate.EMError;
import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMUserInfo;
import com.hyphenate.exceptions.HyphenateException;
import com.hyphenate.wrapper.util.EMSDKMethod;
import com.hyphenate.wrapper.callback.EMCommonValueCallback;
import com.hyphenate.wrapper.callback.EMWrapperCallback;
import com.hyphenate.wrapper.helper.EMUserInfoHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Map;

public class EMUserInfoManagerWrapper extends EMBaseWrapper{
    EMUserInfoManagerWrapper() {
        registerEaseListener();
    }

    public String onMethodCall(String method, JSONObject jsonObject, EMWrapperCallback callback) throws JSONException {
        String ret = null;
        if (EMSDKMethod.updateOwnUserInfo.equals(method)) {
            ret = updateOwnUserInfo(jsonObject, callback);
        } else if (EMSDKMethod.updateOwnUserInfoWithType.equals(method)) {
            ret = updateOwnUserInfoWithType(jsonObject, callback);
        }else if (EMSDKMethod.fetchUserInfoById.equals(method)) {
            ret = fetchUserInfoByUserId(jsonObject, callback);
        }else if (EMSDKMethod.fetchUserInfoByIdWithType.equals(method)) {
            ret = fetchUserInfoByIdWithType(jsonObject, callback);
        } else {
            ret = super.onMethodCall(method, jsonObject, callback);
        }
        return ret;
    }

    private String updateOwnUserInfo(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String username = EMClient.getInstance().getCurrentUser();
        if (username == null) {
            HyphenateException e = new HyphenateException(EMError.USER_NOT_LOGIN,"User not login");
            onError(e, callback);
            return null;
        }

        EMUserInfo userInfo = EMUserInfoHelper.fromJson(params.getJSONObject("userInfo"));
        userInfo.setUserId(username);
        asyncRunnable(() -> EMClient.getInstance().userInfoManager().updateOwnInfo(userInfo, new EMCommonValueCallback<>(callback)));
        return null;
    }


    private String updateOwnUserInfoWithType(JSONObject params, EMWrapperCallback callback) throws JSONException {
        int userInfoTypeInt = params.getInt("userInfoType");
        String userInfoTypeValue = params.getString("userInfoValue");
        EMUserInfo.EMUserInfoType userInfoType = getUserInfoTypeFromInt(userInfoTypeInt);

        asyncRunnable(()-> EMClient.getInstance().userInfoManager().updateOwnInfoByAttribute(userInfoType,userInfoTypeValue,new EMCommonValueCallback<>(callback)));
        return null;
    }


    private String fetchUserInfoByUserId(JSONObject params, EMWrapperCallback callback) throws JSONException {
        JSONArray userIdArray = params.getJSONArray("userIds");
        String[] userIds = new String[userIdArray.length()];
        for (int i = 0; i < userIdArray.length(); i++) {
            userIds[i] = (String) userIdArray.get(i);
        }

        asyncRunnable(()->{
            EMCommonValueCallback<Map<String,EMUserInfo>> callBack = new EMCommonValueCallback<Map<String,EMUserInfo>>(callback) {
                @Override
                public void onSuccess(Map<String,EMUserInfo> object) {
                    JSONObject jo = null;
                    try {
                        jo = generateMapFromMap(object);
                    }catch (JSONException e) {
                        e.printStackTrace();
                    }finally {
                        updateObject(jo);
                    }
                }
            };

            EMClient.getInstance().userInfoManager().fetchUserInfoByUserId(userIds, callBack);
        });
        return null;
    }


    private String fetchUserInfoByIdWithType(JSONObject params, EMWrapperCallback callback) throws JSONException {
        JSONArray userIdArray = params.getJSONArray("userIds");
        JSONArray userInfoTypeArray = params.getJSONArray("userInfoTypes");

        String[] userIds = new String[userIdArray.length()];
        for (int i = 0; i < userIdArray.length(); i++) {
            userIds[i] = (String) userIdArray.get(i);
        }

        EMUserInfo.EMUserInfoType[] infoTypes = new EMUserInfo.EMUserInfoType[userInfoTypeArray.length()];
        for (int i = 0; i < userInfoTypeArray.length(); i++) {
            EMUserInfo.EMUserInfoType infoType = getUserInfoTypeFromInt((int) userInfoTypeArray.get(i));
            infoTypes[i] = infoType;
        }

        asyncRunnable(()->{
            EMCommonValueCallback<Map<String,EMUserInfo>> callBack = new EMCommonValueCallback<Map<String,EMUserInfo>>(callback) {
                @Override
                public void onSuccess(Map<String,EMUserInfo> object) {
                    JSONObject jo = null;
                    try {
                        jo = generateMapFromMap(object);
                    }catch (JSONException e) {
                        e.printStackTrace();
                    }finally {
                        updateObject(jo);
                    }
                }
            };

            EMClient.getInstance().userInfoManager().fetchUserInfoByAttribute(userIds, infoTypes, callBack);
        });

        return null;
    }


    //组装response map
    JSONObject generateMapFromMap(Map<String, EMUserInfo> aMap) throws JSONException{
        JSONObject jsonObject = new JSONObject();
        for(Map.Entry<String, EMUserInfo> entry : aMap.entrySet()){
            String mapKey = entry.getKey();
            EMUserInfo mapValue = entry.getValue();
            jsonObject.put(mapKey, EMUserInfoHelper.toJson(mapValue));
        }
        return jsonObject;
    }


    //获取用户属性类型
    private EMUserInfo.EMUserInfoType getUserInfoTypeFromInt(int value){
        EMUserInfo.EMUserInfoType infoType;

        switch (value){
            case 0:
            {
                infoType = EMUserInfo.EMUserInfoType.NICKNAME;
            }
            break;

            case 1:
            {
                infoType = EMUserInfo.EMUserInfoType.AVATAR_URL;
            }
            break;

            case 2:
            {
                infoType = EMUserInfo.EMUserInfoType.EMAIL;
            }
            break;

            case 3:
            {
                infoType = EMUserInfo.EMUserInfoType.PHONE;
            }
            break;

            case 4:
            {
                infoType = EMUserInfo.EMUserInfoType.GENDER;
            }
            break;

            case 5:
            {
                infoType = EMUserInfo.EMUserInfoType.SIGN;
            }
            break;

            case 6:
            {
                infoType = EMUserInfo.EMUserInfoType.BIRTH;
            }
            break;

            case 100:
            {
                infoType = EMUserInfo.EMUserInfoType.EXT;
            }
            break;

            default:
                throw new IllegalStateException("Unexpected value: " + value);
        }

        return infoType;
    }
    
    private void registerEaseListener (){}
}

File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_iOS/wrapper/wrapper/managers/EMUserInfoManagerWrapper.m
//
//  EMUserInfoManagerWrapper.m
//  wrapper
//
//  Created by 杜洁鹏 on 2022/11/15.
//

#import "EMUserInfoManagerWrapper.h"
#import "EMUserInfo+Helper.h"

@implementation EMUserInfoManagerWrapper
- (instancetype)init {
    if (self = [super init]) {
        [self registerEaseListener];
    }
    
    return self;
}

- (NSString *)onMethodCall:(NSString *)method
                    params:(NSDictionary *)params
                  callback:(EMWrapperCallback *)callback {
    NSString *ret = nil;
    if ([updateOwnUserInfo isEqualToString:method]) {
        ret = [self updateOwnUserInfo:params callback:callback];
    } else if ([updateOwnUserInfoWithType isEqualToString:method]) {
        ret = [self updateOwnUserInfoWithType:params callback:callback];
    }else if ([fetchUserInfoById isEqualToString:method]) {
        ret = [self fetchUserInfoByUserId:params callback:callback];
    }else if ([fetchUserInfoByIdWithType isEqualToString:method]) {
        ret = [self fetchUserInfoByIdWithType:params callback:callback];
    } else {
        ret = [super onMethodCall:method params:params callback:callback];
    }
    return ret;
}

- (NSString *)updateOwnUserInfo:(NSDictionary *)param
                       callback:(EMWrapperCallback *)callback {
    __weak EMUserInfoManagerWrapper *weakSelf = self;
    NSString *usenrame = EMClient.sharedClient.currentUsername;
    if (usenrame == nil) {
        EMError *error = [EMError errorWithDescription:@"User not login" code:EMErrorUserNotLogin];
        [weakSelf wrapperCallback:callback error:error object:nil];
        return nil;
    }
    
    EMUserInfo *userInfo = [EMUserInfo fromJson:param[@"userInfo"]];
    userInfo.userId = usenrame;
    [EMClient.sharedClient.userInfoManager updateOwnUserInfo:userInfo completion:^(EMUserInfo *aUserInfo, EMError *aError) {
        [weakSelf wrapperCallback:callback error:aError object:[aUserInfo toJson]];
    }];
    
    return nil;
}

- (NSString *)updateOwnUserInfoWithType:(NSDictionary *)param
                               callback:(EMWrapperCallback *)callback {
    __weak EMUserInfoManagerWrapper *weakSelf = self;
    
    int typeValue = [param[@"userInfoType"] intValue];
    EMUserInfoType userInfoType = [self userInfoTypeFromInt:typeValue];
    NSString *userInfoValue = param[@"userInfoValue"];
    
    
    [EMClient.sharedClient.userInfoManager updateOwnUserInfo:userInfoValue withType:userInfoType completion:^(EMUserInfo *aUserInfo, EMError *aError) {
        [weakSelf wrapperCallback:callback error:aError object:[aUserInfo toJson]];
    }];
    
    return nil;
}

- (NSString *)fetchUserInfoByUserId:(NSDictionary *)param
                           callback:(EMWrapperCallback *)callback {
    __weak EMUserInfoManagerWrapper *weakSelf = self;
    NSArray *userIds = param[@"userIds"];
    
    [EMClient.sharedClient.userInfoManager fetchUserInfoById:userIds completion:^(NSDictionary *aUserDatas, EMError *aError) {
        
        NSMutableDictionary *dic = NSMutableDictionary.new;
        [aUserDatas enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
            dic[key] = [(EMUserInfo *)obj toJson];
        }];
        
        [weakSelf wrapperCallback:callback error:aError object:dic];
    }];
    return nil;
}

- (NSString *)fetchUserInfoByIdWithType:(NSDictionary *)param
                               callback:(EMWrapperCallback *)callback {
    __weak EMUserInfoManagerWrapper *weakSelf = self;
    NSArray *userIds = param[@"userIds"];
    NSArray<NSNumber *> *userInfoTypes = param[@"userInfoTypes"];
    
    [EMClient.sharedClient.userInfoManager fetchUserInfoById:userIds type:userInfoTypes completion:^(NSDictionary *aUserDatas, EMError *aError) {
        
        NSMutableDictionary *dic = NSMutableDictionary.new;
        [aUserDatas enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
            dic[key] = [(EMUserInfo *)obj toJson];
        }];
        [weakSelf wrapperCallback:callback error:aError object:dic];
    }];
    return nil;
}

- (EMUserInfoType)userInfoTypeFromInt:(int)typeValue {
    EMUserInfoType userInfoType;
    
    switch (typeValue) {
        case 0:
            userInfoType = EMUserInfoTypeNickName;
            break;
        case 1:
            userInfoType = EMUserInfoTypeAvatarURL;
            break;
        case 2:
            userInfoType = EMUserInfoTypePhone;
            break;
        case 3:
            userInfoType = EMUserInfoTypeMail;
            break;
        case 4:
            userInfoType = EMUserInfoTypeGender;
            break;
        case 5:
            userInfoType = EMUserInfoTypeSign;
            break;
        case 6:
            userInfoType = EMUserInfoTypeBirth;
            break;
        case 7:
            userInfoType = EMUserInfoTypeExt;
            break;
        default:
            userInfoType = EMUserInfoTypeNickName;
            break;
    }
    
    return userInfoType;
}

- (void)registerEaseListener{
    
}
@end
